Improve online_delete configuration and DB tuning:

ximinez opened the pull request:
* Document delete_batch, back_off_milliseconds, age_threshold_seconds.
* Convert those time values to chrono types.
* Fix bug that ignored age_threshold_seconds.
* Add a "recovery buffer" to the config that gives the node a chance to
  recover before aborting online delete.
* Add begin/end log messages around the SQL queries.
* Add a new configuration section: [sqlite] to allow tuning the sqlite
  database operations. Ignored on full/large history servers.
* Update documentation of [node_db] and [sqlite] in the
  rippled-example.cfg file.
* Resolves ripple#3321

I added the "recovery buffer" function that wasn't described in the original issue to scratch my own itch, because my local machine was losing sync for a fraction of a second while the ledger was being copied over. This resolved that issue. I defaulted it to not run, so it doesn't cause unexpected behavior.

Note this includes two commits, one from @mtrippled . Both need to be reviewed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ripple/rippled/3429)
<!-- Reviewable:end -->

Commit:
Make server_info report consistent with internal
evaluations for validated ledger age.

ximinez committed:
Improve online_delete configuration and DB tuning:

* Document delete_batch, back_off_milliseconds, age_threshold_seconds.
* Convert those time values to chrono types.
* Fix bug that ignored age_threshold_seconds.
* Add a "recovery buffer" to the config that gives the node a chance to
  recover before aborting online delete.
* Add begin/end log messages around the SQL queries.
* Add a new configuration section: [sqlite] to allow tuning the sqlite
  database operations. Ignored on full/large history servers.
* Update documentation of [node_db] and [sqlite] in the
  rippled-example.cfg file.
* Resolves #3321

MarkusTeufelberger commented on cfg/rippled-example.cfg:
```diff
@@ -1212,23 +1314,24 @@ medium
-online_delete=2000
+type=NuDB
+path=/var/lib/rippled/db/nudb
+online_delete=512
```
If such a small value is the default, please consider also adding a shard_db section with a limit of a few dozen/hundred GiB by default.

ximinez replied:
@MarkusTeufelberger Thanks for that feedback. Those default settings are definitely not yet set in stone. My understanding that the suggestion for the smaller default was to make it easier for more people to participate without requiring huge resources. (The original ticket didn't have a lot of detail.) Of course, there's a balance to be found, because we already have pretty high hardware requirements, and we do want as many nodes as possible to contribute to history storage as a way to help the network.

I'm not sure if shards are ready to be set by default. I'll defer to @miguelportilla on that question.

MarkusTeufelberger commented on cfg/rippled-example.cfg:
```diff
@@ -1212,23 +1314,24 @@ medium
-# delete old ledgers while maintaining at least 2000. Do not require an
-# external administrative command to initiate deletion.
+# found at https://xrpl.org/capacity-planning.html#node-db-type
+# type=NuDB is recommended for non-validators with fast SSDs. Validators or
```
I would use a `none` database backend on a validator. They don't need to concern themselves with historic data at all.

ximinez replied:
The linked capacity planning page recommends RocksDB, but I can definitely see your point about using `none` (or `memory`). My biggest concern with using those options on the individual node level, is that it means it will take much longer to start and restart, because the node will have to download the entire ledger every time.  On the global level, I'm thinking about a failsafe / worst case situation. For example, if every validator shuts down simultaneously :scream:, they will need to have some data on disk so they can bootstrap restarting the network.

MarkusTeufelberger replied:
Anecdotally, my full history node takes about half an hour to start with `--load` vs. a few minutes at most to get synced with `--net`.

Bootstrapping can be done safely with `--ledgerfile` (https://xrpl.org/commandline-usage.html#initial-ledger-options) and that's probably better anyways a better way forward than trusting that data was stored to disk correctly in such a catastrophic scenario. The file itself could be hashed and communicated out-of-band in a lot of different ways compared to a node database that's not even a full shard in most cases.

nbougalis replied:
@MarkusTeufelberger: re: slow start times with `--load`: are you using SSDs?

MarkusTeufelberger replied:
Of course. Over a dozen TB of 'em.

nbougalis replied:
I think that `--net` should become the default now, unless a different command line option (for which I don't have a good name) is specified.

ximinez replied:
@nbougalis I think making `--net` the default is out of scope for this PR. Do you agree?

ximinez committed:
[FOLD] Small cleanup that I missed when preparing the PR

ximinez committed:
[FOLD] Clarify the description of 'recovery_buffer_seconds'

nbougalis commented on cfg/rippled-example.cfg:
```diff
@@ -1212,23 +1317,24 @@ medium
 # of the ripple network that rippled operators shard as much as practical.
-# NuDB requires SSD storage. Helpful information can be found here
+# NuDB requires SSD storage. Helpful information can be found at
 # https://ripple.com/build/history-sharding
```
Please update this URL

ximinez replied:
Good catch. Fixed this and several other ripple.com URLs.

nbougalis commented on src/ripple/core/impl/Config.cpp:
```diff
@@ -454,7 +454,7 @@ Config::loadFromString(std::string const& fileContents)
             FETCH_DEPTH = 0;
         else if (boost::iequals(strTemp, "full"))
-            FETCH_DEPTH = 1000000000u;
+            FETCH_DEPTH = 1'000'000'000u;
```
Consider [`std::numeric_limits<std::uint32_t>::max`](https://en.cppreference.com/w/cpp/types/numeric_limits/max) instead for `FETCH_DEPTH` and `LEDGER_HISTORY`.

ximinez replied:
Do you know if there was a reason this was defined as 1 billion instead of `max()` in the first place?

nbougalis replied:
There was no particular reason.

nbougalis commented on cfg/rippled-example.cfg:
```diff
@@ -892,6 +939,12 @@
 #   [import_db]     Settings for performing a one-time import (optional)
 #   [database_path]   Path to the book-keeping databases.
 #
+#   There are 4 or 5 bookkeeping SQLite database that the server creates and
```
Typo: database -> databases.

ximinez replied:
Fixed.

ximinez committed:
[FOLD] Remove unused catch variable

ximinez committed:
[FOLD] Review feedback from @nbougalis:

(Include these items in the main commit when squashing.)
* Update rippled URLs in documentation.
* Full history and fetch depth use max() value instead of an arbitrary
  value of 1 billion.

nbougalis commented on src/ripple/core/DatabaseCon.h:
```diff
@@ -89,12 +89,17 @@ class DatabaseCon
     DatabaseCon(
         Setup const& setup,
         std::string const& DBName,
+        bool useCommonPragma,
```
I don't know if I like this; I'm generally allergic to functions that accept a `bool` and use it to do one or the other thing. Why can't we set the right values in `setup` directly?

ximinez replied:
It would be pretty simple to set the values in the `Setup` object, now that you mention it. I might need to define a member function or two, just to make the internals easier, and to make it easier switch between the options for `Setup` objects that are used to create multiple `DatabaseCon`s

ximinez replied:
PS. I went down this path because I initially thought the other ctor override would need to use the "common pragma" but without a `Setup` object being passed in. I later realized none of them do, but didn't backtrack to see if the `bool` was still a good idea. So thanks for pointing this out!

nbougalis commented on src/ripple/core/DatabaseCon.h:
```diff
@@ -89,12 +89,17 @@ class DatabaseCon
+        static std::unique_ptr<std::vector<std::string> const> CommonPragma;
+        /// Shortcut used by the database connections that ignore the common
+        /// pragma strings
+        static const std::vector<std::string> NoCommonPragma;
```
μNit: weird case for the name(s) here. Also confused as to why two separate things are needed.

ximinez replied:
A lot of other static class variables have capitalized names, but I don't necessarily have to keep doing it.

nbougalis commented on src/ripple/app/main/DBInit.h:
```diff
@@ -26,13 +26,23 @@ namespace ripple {
+// full history nodes. This is because such a large amount of data will
+// be more difficult to recover if a rare failure occurs, which are
+// more likely with some of the other available tuning settings.
+inline constexpr std::uint32_t SQLITE_TUNING_CUTOFF = 100'000'000;
```
I understand the rationale behind this, but I'm not sure if it's a good idea. I think that it's reasonable to warn in the configuration file and even on every startup about these options, but making the options entirely unavailable is not. Ironically, the existing logic that enforces this will prevent choosing options that are safer than the default even if the server operator _really_ wants them.

I'll give this some more thought, but right now, it seems like a bad idea.

ximinez replied:
I can change it to warning log message(s) pretty easily, and update the config file.

nbougalis reviewed (commented):
Left some comments. Still thinking about this.

ximinez committed:
[fold] review feeback from @nbougalis 2:

* changed the sqlite ledger_history cutoff to log a warning instead of
  ignoring sqlite settings. updated documentation.
* integrated the commonpragma back in to the setup struct and restored
  the interface of the databasecon ctors.

ximinez committed:
[fold] add unit tests for log messages added last commit

* involved moving some test logging subclasses around to be more easily
  accessible, and some minor refactoring of Env.

mtrippled commented on src/ripple/app/misc/SHAMapStoreImp.cpp:
```diff
@@ -666,16 +705,32 @@ SHAMapStoreImp::health()
+
+            age = ledgerMaster_->getValidatedLedgerAge();
+            mode = netOPs_->getOperatingMode();
+        }
```
Notwithstanding recovery_buffer_seconds, the online_delete activity is triggered each time a new ledger is validated. And desyncs seem to usually happen for an entire consensus interval, of ~3-4 seconds. This really seems superfluous, especially adding a new config option to tune this.

ximinez replied:
I'm not married to it, but on my local node, it was taking so long for the ledger to copy that the node would switch to `syncing` before it could finish, and thus it would abort only to almost immediately start rotating again. The thing is that as far as I could tell, it wasn't losing sync because of the ledger copy, but because it build a different ledger during consensus than the UNL validators. Here's an example that includes some log messages that I added only to figure out what was going on:

```
2020-May-27 19:55:17.7525561 UTC SHAMapStore:TRC Nothing to delete from DELETE FROM AccountTransactions WHERE LedgerSeq < %u;
2020-May-27 19:55:17.7547395 UTC SHAMapStore:TRC copying ledger 55752529
2020-May-27 19:55:35.0709935 UTC NetworkOPs:NFO consensusViewChange
2020-May-27 19:55:35.0718717 UTC NetworkOPs:NFO STATE->syncing
2020-May-27 19:55:35.1798535 UTC SHAMapStore:WRN Not deleting. state: syncing. age 1s
2020-May-27 19:55:35.1803785 UTC SHAMapStore:WRN Not deleting. state: syncing. age 1s
2020-May-27 19:55:37.7108585 UTC SHAMapStore:WRN Not deleting. state: syncing. age 0s
2020-May-27 19:55:38.6848155 UTC NetworkOPs:NFO STATE->tracking
2020-May-27 19:55:38.6850136 UTC NetworkOPs:NFO STATE->full
2020-May-27 19:55:41.6406001 UTC SHAMapStore:WRN rotating  validatedSeq 55752535 lastRotated_ 55461357 deleteInterval 30000 canDelete_ 55461356 state full age 0s
```
Note the timestamps. At 19:55:17, it starts copying the ledger. 28 seconds later, the consensus view changes and it drops to syncing. 0.1 seconds later, the ledger copy aborts in `health()`. About 3 seconds after that (the next ledger, I presume), we're back to tracking, then full. 3 seconds later, we start rotating again. My node was doing this for _hours_. It's a beefy machine, but a Windows debug build.

The reason I added it as a config option is so that it _wouldn't_ do this by default. 

The obvious question is whether it's worth adding a feature to support marginal nodes. It seemed like a good idea at the time, and it at least helped me get this feature developed and finished.

I can think of two options:
* Remove it entirely.
* Make the feature `#if DEBUG`, remove the config option, and use a hard-coded value.

I'll go with whatever you think makes more sense.

mtrippled replied:
It could be useful in a troubleshooting context--since you've already done the work, may as well keep it, as an option, but not by default. It's not plausible from that vantage point to determine whether the node itself is flaky, or if online_delete triggered the desync. So, to be safe, assume the latter, which means the default behavior should be to not have a buffer of time to wait for the node to recover.

ximinez replied:
To confirm, the default time is to not have the buffer by default. :+1:

mtrippled commented on src/ripple/app/misc/SHAMapStoreImp.cpp:
```diff
@@ -566,19 +573,34 @@ SHAMapStoreImp::clearSql(
+        // Micro-optimization mainly to clarify logs
+        JLOG(journal_.trace()) << "Nothing to delete from " << deleteQuery;
+        return true;
+    }
```
Actually, we want to delete up to or equal to lastRotated.
For example, the loop on line 609 should be while (min <= lastRotated)
And if min == lastRotated, then break out of the loop/return after the query is executed. Though there may be something more elegant.

ximinez replied:
Are you sure about that? Every other check runs up to, but doesn't include `lastRotated`, and it's documented that way. https://github.com/ripple/rippled/blob/9173a6a7d3a08be4a7e8345bd763802693a8dcdb/src/ripple/app/misc/SHAMapStoreImp.cpp#L360__

mtrippled replied:
To be precise, the code should actually execute the query as it is for all values <=  lastRotated
            DELETE FROM <table> WHERE LedgerSeq < <lastRotated>;

The existing code actually does not call the query when min==lastRotated. It should execute the query. Consider that lastRotated is 1000000. We want to delete everything from 1..999999. But imagine that the function is called initially with min set to 1000000. The delete query will not be called. This is in the existing code. Your opimization made it apparent to me upon close inspection.

Once everything < lastRotated has been deleted, perform no more queries. With the existing code, it would mean checking if min==lastRotated after the delete query, then breaking out of the while loop, and checking for min <= lastRotated in the while condition.

Alternately, something like this:

    while (min <= lastRotated)
    {
        LedgerIndex minRange = std::min(lastRotated, min + deleteBatch);
        {
            auto db =  database.checkoutDb ();
            *db << boost::str (formattedDeleteQuery % minRange);
        }
        if (health())
            return true;
        min = minRange;
        if (min < lastRotated)
            std::this_thread::sleep_for (
                    std::chrono::milliseconds (backOff_));
    }

ximinez replied:
But `min` is initialized from the data in the table.
```
    LedgerIndex min = std::numeric_limits<LedgerIndex>::max();

    {
        boost::optional<std::uint64_t> m;
        JLOG(journal_.trace())
            << "Begin: Look up lowest value of: " << minQuery;
        {
            auto db = database.checkoutDb();
            *db << minQuery, soci::into(m);
        }
        JLOG(journal_.trace()) << "End: Look up lowest value of: " << minQuery;
        if (!m)
            return;
        min = *m;
    }
```
So if `lastRotated` is 1,000,000, and the `minQuery` returns `1,000,000`, that means that there are no records in that particular table from 1..999,999. That's why I added the extra check, and why the loop doesn't need to run. We _know_ that the DELETE query won't delete any rows, because we just discovered that there are no rows to delete.

mtrippled replied:
You're correct. The change I suggested doesn't need to be made. Also, keeping your original change is OK.

ximinez commented on src/ripple/core/impl/DatabaseCon.cpp:
```diff
@@ -38,9 +40,119 @@ setup_DatabaseCon(Config const& c)
+            result->reserve(3);
+
+            // defaults
+            std::string safety_level = "high";
```
I disagree, but I think it's because of how I'm viewing the settings overall. I think of the `safety_level` as a baseline of settings, then all the others as overrides of the `safety_level`. In other words, the default is "high", but if I want transactions to be asyncronous, I can set `synchronous=off`. Similarly, if I want the lower safety level, but want to make sure that logging is safer, I could set
```
safety_level=low
journal_mode=wal
```

It's entirely possible I'm overthinking this. I could update the documentation in the example config file to be clearer that safety_level is the baseline and the others override.

> I think part of what you're doing with showWarning addresses this

The intention of `showWarning` is specific to when the config is set up to store a _lot_ of history. Think full history node, but since it's just a warning message, I set the cutoff lower.

ximinez replied:
> Why would anybody choose only 1 or 2, and not all 3?

Excellent question. Some of the options can actually lower the data integrity risk, at the cost of extra I/O. I figured if I'm going to provide the options, I should go all the way, since I'm sure there are some folks who would use them.

> Perhaps safety_level should be removed

The more I think about it, the more I think it makes sense to make `safety_level` mutually exclusive from the other options. Someone who wants a simple option can use `safety_level=low`, and someone who wants to tune with more detail can whatever combos of the other three that they want. In other words, the example I gave above is starting to feel contrived and unrealistic.

mtrippled replied:
I agree.

ximinez commented on src/ripple/app/main/Main.cpp:
```diff
@@ -555,7 +556,9 @@ run(int argc, char** argv)
             session << "PRAGMA temp_store_directory=\"" << tmpPath.string()
                     << "\";";
             session << "VACUUM;";
-            session << "PRAGMA journal_mode=WAL;";
```
Turns out that `temp_store_directory` is deprecated: https://www.sqlite.org/pragma.html#pragma_temp_store_directory.

It looks like you're right about the `journal_mode`: https://www.sqlite.org/lang_vacuum.html
> The VACUUM command works by copying the contents of the database into a temporary database file and then overwriting the original with the contents of the temporary file. When overwriting the original, a rollback journal or write-ahead log WAL file is used just as it would be for any other database transaction.

So I'll update that.

ximinez committed:
[FOLD] Review feedback from @mtravis, part 1:

* Make `safety_level` and other tuning options mutually exclusive
* Clean up VACUUM functionality and tuning to use settings.
* Fix bugs in the config setup where it still ignored tuning settings if
  full history. Add tests to catch that mistake.
* Fix some bugs in the unit tests, and make them more robustly check
  error messages.

ximinez commented on src/ripple/app/misc/SHAMapStoreImp.cpp:
```diff
@@ -348,23 +359,14 @@ SHAMapStoreImp::run()
         // will delete up to (not including) lastRotated
         if (validatedSeq >= lastRotated + deleteInterval_ &&
-            canDelete_ >= lastRotated - 1)
+            canDelete_ >= lastRotated - 1 && !health())
```
Primarily, it's to prevent writing the log message if no actual work is going to get done. Compare to what it was doing before: log the "rotating" message, then do a `health()` check, then `clearPrior()`.

If any change should be made, I'd think it would be to take the first `health()` check out of `clearPrior()`. What do you think?

mtrippled replied:
Yes, removing the first health() from clearPrior() makes sense.

ximinez replied:
Done

ximinez commented on src/ripple/app/misc/SHAMapStoreImp.cpp:
```diff
@@ -566,19 +573,34 @@ SHAMapStoreImp::clearSql(
+            assert(false);
+            return false;
+        }
+        JLOG(journal_.trace()) << "End: Look up lowest value of: " << minQuery;
```
I honestly don't remember for sure why I added the `if (auto db = database.checkoutDb())` check... I think it was so that it would go out of scope and release the lock sooner, and then I added the `else` because it "wouldn't hurt". It would make more sense to rewrite it as a simple local block.

ximinez commented on src/ripple/app/misc/SHAMapStoreImp.cpp:
```diff
@@ -566,19 +573,34 @@ SHAMapStoreImp::clearSql(
+        else
+        {
+            assert(false);
+            return false;
```
Good catch! Will fix.

ximinez committed:
[FOLD] Review feedback from @mtravis, part 2:

* Change some log message levels.
* clearSql no longer returns a bool that every caller ignores.
* Restructure a few unnecessary validity checks

ximinez committed:
[FOLD] Remove an unnecessary health check

seelabs commented on src/ripple/app/main/DBInit.h:
```diff
@@ -63,15 +73,14 @@ inline constexpr auto TxDBName{"transaction.db"};
     {
 #else
-    std::array<char const*, 5> TxDBPragma {{
+    std::array<char const*, 3> TxDBPragma {{
```
The preprocessor guards make this declaration hard to read. Now that we're on c++-17 I think we can simplify this to:
```

inline constexpr std::array TxDBPragma
{
    "PRAGMA page_size=4096;", "PRAGMA journal_size_limit=1582080;",
        "PRAGMA max_page_count=2147483646;",
#if (ULONG_MAX > UINT_MAX) && !defined(NO_SQLITE_MMAP)
        "PRAGMA mmap_size=17179869184;",
#endif
};
```

ximinez replied:
Nice. Done.

seelabs commented on src/ripple/app/main/Main.cpp:
```diff
@@ -522,23 +519,24 @@ run(int argc, char** argv)
-                          << " bytes.\n";
-                return -1;
+                auto available = space(dbPath.parent_path()).available;
+                if (available < dbSize)
```
Could move the decl into the `if` and lose the extra scope:
```
            if (auto available = space(dbPath.parent_path()).available;
                available < dbSize)
```

seelabs commented on src/ripple/app/misc/SHAMapStoreImp.h:
```diff
@@ -106,8 +107,9 @@ class SHAMapStoreImp : public SHAMapStore
-    std::int32_t ageThreshold_ = 60;
+    std::chrono::milliseconds backOff_{100};
+    std::chrono::seconds ageThreshold_{60};
+    boost::optional<std::chrono::seconds> recoveryBuffer_{};
```
When I see something called `xxxBuffer` I think of memory buffers. We're using "buffer" in a different sense here. Maybe cal lit "Margin"? Or something else?

We should also leave a brief comment explaining what this does. I know the others don't, but it would be good to add comment for the new fields at least.
Also, nit: No need for the extra `{}`

ximinez replied:
`recoveryWaitTime_` and renamed the config setting to `recovery_wait_seconds`.

seelabs commented on src/ripple/app/misc/SHAMapStoreImp.h:
```diff
@@ -218,7 +220,7 @@ class SHAMapStoreImp : public SHAMapStore
      *  @return true if any deletable rows were found (though not
```
Outdated comment

seelabs commented on src/ripple/app/misc/SHAMapStoreImp.cpp:
```diff
@@ -587,17 +606,25 @@ SHAMapStoreImp::clearSql(
     }
     JLOG(journal_.debug()) << "finished: " << deleteQuery;
-    return true;
+    return;
```
Could remove this return.

seelabs commented on src/ripple/app/misc/SHAMapStoreImp.cpp:
```diff
@@ -666,16 +694,32 @@ SHAMapStoreImp::health()
+                << "s for node to get back into sync with network. state: "
+                << app_.getOPs().strOperatingMode(mode, false) << ". age "
+                << age.count() << 's';
+            std::this_thread::sleep_for(*recoveryBuffer_);
```
Sleeping in a call to "health" is a bit of a surprising interface. I can see why you did this. Since `health` is always called from functions on the main `run` thread, this should be OK. Still, we should probably add a comment to "health" saying the function can sleep and that it should never be called from a function outside of the `run` thread including functions that part of the public interface.

seelabs commented on src/ripple/core/impl/DatabaseCon.cpp:
```diff
@@ -38,9 +40,134 @@ setup_DatabaseCon(Config const& c)
+            if (set(safety_level, "safety_level", sqlite))
+            {
+                showRiskWarning = boost::iequals(safety_level, "low");
+                if (showRiskWarning)
```
I'd rewrite this to make the condition `boost::iequals(safety_level, "low")` and then setting `showRiskWarning` in the if statement. That way it's clearer that we're running that code if the `safety_level` is low, it's only incidentally related to `showRiskWarning`.

ximinez replied:
That's how I had it the first time, but I changed it to match the pattern on the other items. But it's not quite the same pattern, and it does make sense to do it your way here.

seelabs commented on src/ripple/nodestore/impl/Shard.cpp:
```diff
@@ -683,6 +684,7 @@ Shard::initSQLite(std::lock_guard<std::recursive_mutex> const&)
 
         if (backendComplete_)
         {
+            setup.noPragma();
```
Maybe initialize `setup` so it's `const` and not change the `setup` object in the arms of the if statement? Something like:
```
    DatabaseCon::Setup const setup = [&] {
        DatabaseCon::Setup result;
        result.startUp = config.START_UP;
        result.standAlone = config.standalone();
        result.dataDir = dir_;
        if (backendComplete_)
            result.noPragma();
        else
            result.usePragma();
        return result;
    }();
```

ximinez replied:
Done

seelabs commented on src/test/jtx/CheckMessageLogs.h:
```diff
@@ -0,0 +1,74 @@
+        @param msg The message string to search for
+        @param found The variable to set to true if the message is found
+    */
+    CheckMessageLogs(std::string msg, bool& found)
```
I think the interface is clearer if we use raw pointers for this interface. I.e. the caller would look like:
```
CheckMessageLogs("MISMATCH ", &found);
```
If you didn't write the test code, I think it's easier to guess what `CheckMessageLogs` does if we pass by pointer.

ximinez replied:
Makes sense. Done.

seelabs reviewed (commented):
Left some minor comments, but overall this looks good.

mtrippled commented on src/ripple/app/main/Main.cpp:
```diff
@@ -552,9 +549,6 @@ run(int argc, char** argv)
                       << std::endl;
 
-            session << "PRAGMA journal_mode=OFF;";
-            session << "PRAGMA temp_store_directory=\"" << tmpPath.string()
```
The purpose of the temp_store_directory is to make sure there's enough space to perform a VACUUM, which essentially requires as much as the entire database being VACUUMed. The temp_store in memory will likely cause a system to run out of RAM. Imagine transaction DB with 2TB+. (https://sqlite.org/tempfiles.html)

ximinez replied:
Good catch! I'll change that.

ximinez commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -2757,16 +2757,24 @@ NetworkOPsImp::getServerInfo(bool human, bool admin, bool counters)
-            auto lCloseTime = lpClosed->info().closeTime;
-            auto closeTime = app_.timeKeeper().closeTime();
-            if (lCloseTime <= closeTime)
+            constexpr std::chrono::seconds HIGH_AGE_THRESHOLD{1000000};
```
Fixed

ximinez commented on cfg/rippled-example.cfg:
```diff
@@ -907,12 +960,77 @@
+#       safety_level        Valid values: high, low
+#                           The default is "high", and tunes the SQLite
+#                           databases in the most reliable mode. "low"
+#                           is equivalent to
```
Good suggestion. Done.

ximinez commented on cfg/rippled-example.cfg:
```diff
@@ -907,12 +960,77 @@
+#                           This setting may not be combined with the
+#                           "safety_level" setting.
+#
+#       temp_store          Valid values: default, file, memory
```
I'm honestly not sure. It came up in my research as potentially useful, but I couldn't really find any examples of rippled using temporary objects. However, that doesn't mean that it doesn't, and I figured that if someone is willing to take the risk, it doesn't hurt to give them the option. I'm weakly in favor of adding/keeping it unless we can definitively prove that it doesn't matter.

seelabs replied:
Ok let's keep it.

ximinez commented on src/ripple/core/DatabaseCon.h:
```diff
@@ -139,11 +156,20 @@ class DatabaseCon
     void
     init(
```
The only reason I can see is that the `pPath` is a little ugly to do inline. Otherwise, it seems to work fine.

mDuo13 commented on cfg/rippled-example.cfg:
```diff
@@ -892,6 +939,12 @@
 #   [import_db]     Settings for performing a one-time import (optional)
 #   [database_path]   Path to the book-keeping databases.
 #
+#   There are 4 or 5 bookkeeping SQLite databases that the server creates and
```
Why is it "4 or 5"? Is one of them only added with certain configurations? Without more context, the phrase "4 or 5" feels uncertain and vague. Maybe "4 to 5" would carry the right implication.

ximinez replied:
I think `state.db` is created if you're using `online_delete`, so that would be the 5th one. I'll make the change.

mDuo13 commented on cfg/rippled-example.cfg:
```diff
@@ -892,6 +939,12 @@
+#   maintains. If you omit this configuration setting, it will default to
+#   creating a directory called "db" located in the same place as your
+#   rippled.cfg file. Partial pathnames will be considered relative to
+#   the location of the rippled executable.
```
nit: instead of "it will default to creating a directory", just say, "the server creates a directory." Similarly, instead of "will be considered relative" just say "are relative".

ximinez replied:
Changed and reworded to be less passive tone.

mDuo13 commented on cfg/rippled-example.cfg:
```diff
@@ -1212,24 +1330,25 @@ medium
-# external administrative command to initiate deletion.
+# found at https://xrpl.org/capacity-planning.html#node-db-type
+# type=NuDB is recommended for non-validators with fast SSDs. Validators or
+#    slow / spinning disks should use RocksDB.
```
Consider adding a note like this:

> Caution: Spinning disks are not recommended. They do not perform well enough to consistently remain synced to the network.

mDuo13 commented on src/ripple/app/ledger/Ledger.cpp:
```diff
@@ -228,14 +228,14 @@ Ledger::Ledger(
     {
         loaded = false;
-        JLOG(j.warn()) << "Don't have TX root for ledger";
+        JLOG(j.warn()) << "Don't have TX root for ledger" << info_.seq;
```
Nit: as long as we're changing these messages, we could make them clearer by removing the abbreviations "TX" (here) and "AS" (below, line 238). Instead, just say "transaction" and "state data".

mDuo13 reviewed (commented):
I looked over the documentation changes. Definitely a big improvement over the previous version! I left a few optional suggestions to tune up the language in a few places, but :+1: with or without those changes.

ximinez commented on src/ripple/core/DatabaseCon.h:
```diff
@@ -116,7 +133,7 @@ class DatabaseCon
         std::array<char const*, M> const& initSQL)
     {
-        init((dataDir / DBName), pragma, initSQL);
+        init((dataDir / DBName), {}, pragma, initSQL);
```
Fixed

ximinez commented on src/ripple/core/DatabaseCon.h:
```diff
@@ -89,6 +90,22 @@ class DatabaseCon
         boost::filesystem::path dataDir;
+        // If unseated, then the `globalPragma` are not used,
+        // otherwise should point to `globalPragma`
+        std::shared_ptr<std::vector<std::string> const> commonPragma;
```
Yes, I belive I could. This is different than the `bool` param that @nbougalis objected to in https://github.com/ripple/rippled/pull/3429#discussion_r435656253, because it's a member with a clearer purpose. I can also abstract that away so the caller only sees a function call.

ximinez commented on src/ripple/core/DatabaseCon.h:
```diff
@@ -89,6 +90,22 @@ class DatabaseCon
+            commonPragma = globalPragma;
+        }
+
+        static std::shared_ptr<std::vector<std::string> const> globalPragma;
```
Alas. But I'm changing it to a `unique_ptr`.

ximinez committed:
[FOLD] Review feedback from @seelabs and @mtrippled:

* Rename retry_buffer_seconds to retry_wait_seconds.
* Improve documentation and examples in example.cfg file.
* Simplify the DatabaseCon::Setup pragma handling.
* Other misc cleanups

seelabs commented on src/ripple/core/DatabaseCon.h:
```diff
@@ -115,8 +131,8 @@ class DatabaseCon
         std::string const& DBName,
         std::array<char const*, N> const& pragma,
         std::array<char const*, M> const& initSQL)
+        : DatabaseCon({}, dataDir / DBName, {}, pragma, initSQL)
```
See the comment below about removing the first parameter to this function. However, _if_ we wanted to use this idiom, I'd give the full class name `DatabaseCon(Base{}, ...)` to make the code easier to navigate; I'd probably rename the `Base` parameter as well, but it's not worth bike-shedding now since I think we should remove it anyway.

seelabs commented on src/ripple/core/DatabaseCon.h:
```diff
@@ -135,15 +151,28 @@ class DatabaseCon
-    void
-    init(
+    DatabaseCon(
+        Base,
```
I don't understand the motivation for adding the `Base` parameter. The function doesn't appear to be ambiguous without it. Unless I'm missing the reason, can we remove this parameter?

ximinez replied:
Basically, if I continue to use `{}` to specify the `commonPragma` param, then it is ambiguous. However, because of the rewrite in the last commit to use a raw pointer, it is not ambiguous if I use `nullptr`, so I've removed `Base`, and changed the caller to use `nullptr`.

seelabs reviewed (approved):
:+1: I did leave one nit that I think we should resolve (the `Base` parameter on the `DatabaseCon` constructor). However since that's so minor I'll give a thumbs up now.

mtrippled commented on src/ripple/app/main/Main.cpp:
```diff
@@ -517,33 +517,34 @@ run(int argc, char** argv)
+            // Only the most trivial databases will fit in memory on typical
+            // (recommended) software. Force temp files to be written to disk
+            // regardless of the config settings.
+            session << boost::format(CommonDBPragmaTemp) % "file";
```
The original behavior needs to come back--namely, that the actual temp directory be specified. Otherwise, SQLIte will by default use the directory in which the database is.

Alternately, I think that getting rid of this feature altogether makes a lot of sense. If somebody *really* wants to VACUUM a database, it's easier to use the command line tool.

mtrippled reviewed (approved):
LGTM
:+1:

ximinez committed:
[FOLD] Review feedback from @mduo13 and @seelabs:

* Improved some of the documentation and log messages
* Removed the unnecessary `DatabaseCon::Base` ctor helper class

mDuo13 commented:
New changes  :+1:  from me.
